### PR TITLE
OSTComposer StakeRequest object storage optimization

### LINKED ISSUE #770: Stop storing StakeRequest object in OSTComposer contract's storage
Storing `StakeRequest` objects in contract's storage are very expensive. Instead, we can change interfaces of the following external functions: `acceptStakeRequest()`, `revokeStakeRequest()` and `rejectStakeRequest()`, to pass all needed params to calculate hash of a `StakeRequest`.

### PR DESCRIPTION
**Pull request by abhayks1**
- stakeRequests mapping has been updated to
  ```mapping(bytes32 => bool) public stakeRequests```

- StakeRequest objects contract's storage optimization

- Below method interfaces has been updated so that stakeRequestHash can be calculated: 
   - acceptStakeRequest()
   - revokeStakeRequest()
   - rejectStakeRequest()

Please see detailed gas analysis in the ticket.

Fixes #770

### CODE REVIEW

**Comment by puneet-khushwani-eth** [comment]
Reviewing.

File: contracts/test/gateway/TestOSTComposer.sol [hunk 1]
```diff
@@ -71,37 +71,16 @@ contract TestOSTComposer is OSTComposer {
...
-        stakeRequests[_stakeHash] = StakeRequest({
-            amount: _amount,
-            beneficiary: _beneficiary,
-            gasPrice: _gasPrice,
-            gasLimit: _gasLimit,
-            nonce: _nonce,
-            staker: _staker,
-            gateway: _gateway
-        });
+        stakeRequests[_stakeHash] = _value;
```
Comment (puneet-khushwani-eth) [inline contracts/test/gateway/TestOSTComposer.sol]: As per `mapping(bytes32 => bool) public stakeRequests`

should we change it to 

`stakeRequests[_stakeHash] = true;`

File: contracts/test/gateway/TestOSTComposer.sol [hunk 2]
```diff
@@ -71,37 +71,16 @@ contract TestOSTComposer is OSTComposer {
...
-        address _staker,
-        EIP20GatewayInterface _gateway,
-        uint256 _amount,
-        uint256 _gasPrice,
-        uint256 _gasLimit,
-        address _beneficiary,
-        uint256 _nonce,
-        bytes32 _stakeHash
+        bytes32 _stakeHash,
+        bool _value
```
Comment (puneet-khushwani-eth) [inline contracts/test/gateway/TestOSTComposer.sol]: `bool _value` might not be needed if we agree upon the comment below.

Comment (puneet-khushwani-eth) [reply contracts/test/gateway/TestOSTComposer.sol]: On second thoughts, we should keep it to allow us testing with `false` later (if needed)

File: contracts/gateway/OSTComposer.sol [hunk 3]
```diff
@@ -265,103 +223,166 @@ contract OSTComposer is Organized, Mutex, ComposerInterface {
...
-        EIP20GatewayInterface gateway = stakeRequest.gateway;
+        delete stakeRequestHashes[_staker][address(_gateway)];
+        delete stakeRequests[stakeRequestHash];
 
-        StakerProxy stakerProxy = stakerProxies[stakeRequest.staker];
+        StakerProxy stakerProxy = stakerProxies[_staker];
         require(
             address(stakerProxy) != address(0),
             "StakerProxy address is null."
         );
```
[the remark this answers is not in the document]
Comment (abhayks1) [reply contracts/gateway/OSTComposer.sol]: Done!

**Comment by abhayks1** [comment]
Travis is green 📗 Available for review 🐮

**Pull request merged**